# bulk_rna-seq
This is a pipeline I used while at UTSA to determine DEG from bulk RNA-seq experiment data. This is intended as a teaching tool for anyone learning to code and use the super computing cluster at UTSA regularly and contains in-depth step-by-step instructions to get started pulling and analyzing bulk RNA-seq datasets.   

#Hello! Here is an index on how to work with each of these files so you can keep them in order
#Open and read all of these files in this order as you go through your RNA-seq pipeline

[1] Updated_Shamu_login_flow (Non-UTSA students can skip this. You will have to set up your own personal computer or server to run the code on your own though.)
[2] sratoolkit_fastq-dump_tool_documentation (You should read this before using sratoolkit but if you are already familiar with sratoolkit you can skip this as well)
[3] fastq_dump_3-5-2021 (This is all the data I use in the pipeline described here)
[4] Download_ref_genome
[5] Installing_BioConductor_packages_locally (Again, Non-UTSA students can skip this part.)
[6] Rsubread_alignment. This is the first half of the RNA_seq_diff_expD-M-YYYY.R file. Run this on a Server then save the environment and run the rest on your personal computer.
[7] RNA_seq_diff_expD_M_YYYYR. This is the second part of R file run this ON_PERSONAL_COMP using the R environment file in the folder.

#Happy coding
